### DISCUSSION TITLE
Performant persistence of models within component / # DB calls / Model hydration / Serialization / Computed Properties etc.

### DISCUSSION THREAD

**Discussion by booni3** [livewire/livewire#1108] [4. General]
Back in https://github.com/livewire/livewire/issues/488, It was decided to start using the `SerializeModels` trait from the Laravel queue system. This allows us to store a model into a public property and have it re-hydrated between requests.

This is great UX... but I think it adds some significant performance issues to certain components. In something, I have just made a simple counter (that has to check a few DB rows on each count) ended up making 10 DB calls just to increment the counter.

I refactored my code to only store ID's in the public properties and use `computed properties` to get model instances when needed. This reduced my calls from 10 down to 3 for this one action.

In the below example, when you run the count function using `wire:click="count"`, a database query is called to the post table every time.

**Before**
```php
public $post;
public $count = 0;

public function mount(Post $post)
{
    $this->post = $post;
}

public function count()
{
    $this->count ++;
}
```

**after**
```php
public $postId;
public $count = 0;

public function mount(Post $post)
{
    $this->postId = $post->id;
}

public function getPostProperty()
{
   return Post::find($this->postId);
{

public function count()
{
    $this->count ++;
}
```
How can we better control this. The UX of serializing models is great but do we really need to run a DB call for every hydration on every request? Can we cache the hydrated models and skip the DB call?

What is the current thinking around how to achieve a performant system?

**Comment by merthurturk** [comment]
@booni3 I had the same issue and did the same thing, stored the id of the model in a public property and used computed properties. I think this might be the best way to do this. With caching, many other questions arise. For example; when to invalidate the cache? Seems to be a simple question but lots of details and edge cases in it.

Hydrated models bring a good developer experience but with a price. IMO a warning block with a good explanation of advantages and disadvantages in the documentation (https://laravel-livewire.com/docs/rendering-components#route-model-binding) would be enough.

**Comment by booni3** [comment]
The DX of the hydrated models is perfect, but its a real shame to sacrifice performance and go with the computed properties workaround.

Considering the gotcha's within using computed properties and caching anyway, it would we **awesome** if Livewire provided a means to cache computed properties per request in the same way.

I am guessing the main issue with this is that it is not possible to detect when public property is accessed to provide the cache and instead `SerializeModels` runs on every property on every request. You would probably have to use magic methods and it starts to get quite confusing to understand what is going on.
